gcc/config/riscv/rvp.md

```diff
@@ -243,3 +254,147 @@
+  [(set (match_operand:SI 0 "register_operand" "=r")
+	(truncate:SI
+	  (ashiftrt:DI
+	    (plus:DI
+	      (plus:DI
+		(sign_extend:DI (match_operand:SI 1 "register_operand" "r"))
+		(sign_extend:DI (match_operand:SI 2 "register_operand" "r")))
+	      (const_int 1))
+	  (const_int 1))))]
+  "TARGET_ZPN && !TARGET_64BIT"
+  "ave\t%0, %1, %2"
+  [(set_attr "type" "dsp")
+   (set_attr "mode" "SI")])
+
+(define_expand "riscv_avedi"
+  [(match_operand:DI 0 "register_operand" "")
+   (match_operand:DI 1 "register_operand" "")
+   (match_operand:DI 2 "register_operand" "")]
+ "TARGET_ZPN && TARGET_64BIT"
+{
+  emit_insn (gen_avedi (operands[0], operands[1], operands[2]));
+  DONE;
+})
+
+(define_insn "avedi"
+  [(set (match_operand:DI 0 "register_operand" "=r")
+	(truncate:DI
+	  (ashiftrt:TI
+	    (plus:TI
+	      (plus:TI
+		(sign_extend:TI (match_operand:DI 1 "register_operand" "r"))
+		(sign_extend:TI (match_operand:DI 2 "register_operand" "r")))
+	      (const_int 1))
+	  (const_int 1))))]
+  "TARGET_ZPN && TARGET_64BIT"
+  "ave\t%0, %1, %2"
+  [(set_attr "type" "dsp")
+   (set_attr "mode" "DI")])
+
+;; bitrev
+(define_insn "riscv_bitrev"
+  [(set (match_operand:SI 0 "register_operand"             "=r,   r")
+	(unspec:SI [(match_operand:SI 1 "register_operand" " r,   r")
+		    (match_operand:SI 2 "rimm5u_operand"   " r, u05")]
+		   UNSPEC_BITREV))]
+  "TARGET_ZPN && !TARGET_64BIT"
+  "@
+   bitrev\t%0, %1, %2
+   bitrevi\t%0, %1, %2"
+  [(set_attr "type"   "dsp")
+   (set_attr "mode"   "SI")])
+
+(define_insn "riscv_bitrev64"
+  [(set (match_operand:DI 0 "register_operand"             "=r,   r")
+	(unspec:DI [(match_operand:DI 1 "register_operand" " r,   r")
+		    (match_operand:SI 2 "rimm6u_operand"   " r, u06")]
+		   UNSPEC_BITREV))]
+  "TARGET_ZPN && TARGET_64BIT"
+  "@
+   bitrev\t%0, %1, %2
+   bitrevi\t%0, %1, %2"
+  [(set_attr "type"   "dsp")
+   (set_attr "mode"   "SI")])
+
+;; bpick
+(define_insn "riscv_bpick<X:mode>"
+  [(set (match_operand:X 0 "register_operand"       "=r")
+	  (ior:X
+	    (and:X
+	      (match_operand:X 1 "register_operand" " r")
+	      (match_operand:X 3 "register_operand" " r"))
+	    (and:X
+	      (match_operand:X 2 "register_operand" " r")
+	      (not:X (match_dup 3)))))]
+  "TARGET_ZPN"
+  "bpick\t%0, %1, %2, %3"
+  [(set_attr "type"   "dsp")
+   (set_attr "mode"   "<MODE>")])
+
+;; clrov
+(define_insn "riscv_clrov<X:mode>"
+  [(unspec_volatile:X [(const_int 0)] UNSPEC_CLROV)]
+  "TARGET_ZPN"
+  "csrrci zero, vxsat, 1"
+  [(set_attr "mode" "<MODE>")])
+
+;; clrs, clz
+(define_insn "riscv_<rvp_optab><mode>2"
+  [(set (match_operand:VECI 0 "register_operand" "=r")
+        (unop:VECI (match_operand:VECI 1 "register_operand" "r")))]
+  "TARGET_ZPN"
+  "<rvp_insn><bits>\t%0, %1"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "<MODE>")])
+
+(define_insn "riscv_<rvp_optab>si2"
+  [(set (match_operand:SI 0 "register_operand" "=r")
+        (unop:SI (match_operand:SI 1 "register_operand" "r")))]
```

Comment:
I would suggest using `standard pattern name`[1] for those pattern, I mean without the riscv_ prefix, using `standard pattern name` can teach GCC to generate that instruction when user using `__builtin_clz` and `__builtin_clrsb`[2]

[1] https://gcc.gnu.org/onlinedocs/gccint/Standard-Names.html#index-clrsbm2-instruction-pattern
[2] https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fclz

Reply:
Thanks for your info! Fixed in commit 715c9ca. Also, there are many other patterns that can be renamed (remove riscv_ prefix), such as smin, smax, etc. Maybe I can just remove the `riscv_` prefix in all the names?